### PR TITLE
feat(app): Add optional 'name' to Source object

### PR DESCRIPTION
**Pull request by CefBoud**
Closes https://github.com/argoproj/argo-cd/issues/17731

# Description

This PR adds an optional 'name' to the Application Source object. It can be used instead of the source position in multi-source Applications CLI.

# Screenshots

image
image
image

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by nitishfy — COMMENTED** [review]
If this PR is ready for review, can you go ahead and fix the CI checks?

**Comment by CefBoud** [comment]
> ### ❗ Preview Environment deployment failed on [Bunnyshell](https://www.bunnyshell.com)
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/5bDne62xgm) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/0VWg8DZq6K)
> 
> Available commands (reply to this comment):
> 
> * 🚀 `/bns:deploy` to redeploy the environment
> * ❌ `/bns:delete` to remove the environment
/bns:deploy

**Comment by CefBoud** [comment]
> ### ❗ Preview Environment deployment failed on [Bunnyshell](https://www.bunnyshell.com)
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/5bDne62xgm) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/0VWg8DZq6K)
> 
> Available commands (reply to this comment):
> 
> * 🚀 `/bns:deploy` to redeploy the environment
> * ❌ `/bns:delete` to remove the environment



/bns:deploy

**Comment by CefBoud** [comment]
> ### ❗ Preview Environment deployment failed on [Bunnyshell](https://www.bunnyshell.com)
> See: [Environment Details](https://environments.bunnyshell.com/b/environment/5bDne62xgm) | [Pipeline Logs](https://environments.bunnyshell.com/b/pipeline/0VWg8DZq6K)
> 
> Available commands (reply to this comment):
> 
> * 🚀 `/bns:deploy` to redeploy the environment
> * ❌ `/bns:delete` to remove the environment

🚀 /bns:deploy

File: cmd/util/app.go [hunk 1]
```diff
@@ -89,6 +89,7 @@ type AppOptions struct {
 	retryBackoffMaxDuration         time.Duration
 	retryBackoffFactor              int64
 	ref                             string
+	SourceName                      string
```
Comment (agaudreault) [inline cmd/util/app.go:92]: All other properties are lowercase.
```suggestion
	sourceName                      string
```

File: test/e2e/app_multiple_sources_test.go [hunk 2]
```diff
@@ -167,3 +167,62 @@ func TestMultiSourceAppWithSourceOverride(t *testing.T) {
...
+		Expect(Success("")).
+		Given().Timeout(60).
+		When().Wait().Then().
+		Expect(Success("")).
+		And(func(app *Application) {
+			statusByName := map[string]SyncStatusCode{}
+			for _, r := range app.Status.Resources {
+				statusByName[r.Name] = r.Status
+			}
+			// check if the app has 3 resources, guestbook and 2 pods
```
Comment (agaudreault) [inline test/e2e/app_multiple_sources_test.go:222]: I think you should validate that the final source is equal to two and has "dynamic duo" and "guestbook2".

**Review by agaudreault — CHANGES_REQUESTED** [review]
Good work 👍 The code changes are missing unit tests and maybe e2e too on the different commands. @ishitasequeira has more knowledge on the usage of multi-source. Do you see something that has been forgotten in this PR?

File: cmd/argocd/commands/app.go [hunk 3]
```diff
@@ -381,6 +396,19 @@ func NewApplicationGetCommand(clientOpts *argocdclient.ClientOptions) *cobra.Com
 			})
 			errors.CheckError(err)
 
+			if sourceName != "" && sourcePosition != -1 {
```
Comment (ishitasequeira) [inline cmd/argocd/commands/app.go:399]: Since the `Name` parameter has been added to the `ApplicationSource` object, it's not exclusive to multi-source applications. This means the `name` field also applies to single-source apps. Therefore, in all commands where we’re adding the `sourceName` check, we should ensure the command supports single-source applications as well, using the `sourceName` field.

File: cmd/argocd/commands/app.go [hunk 4]
```diff
@@ -435,6 +463,7 @@ func NewApplicationGetCommand(clientOpts *argocdclient.ClientOptions) *cobra.Com
 	command.Flags().BoolVar(&hardRefresh, "hard-refresh", false, "Refresh application data as well as target manifests cache")
 	command.Flags().StringVarP(&appNamespace, "app-namespace", "N", "", "Only get application from namespace")
 	command.Flags().IntVar(&sourcePosition, "source-position", -1, "Position of the source from the list of sources of the app. Counting starts at 1.")
+	command.Flags().StringVar(&sourceName, "source-name", "", "Name of the source from the list of sources of the app.")
```
Comment (ishitasequeira) [inline cmd/argocd/commands/app.go:466]: Same here, considering `Name` is not specific to a multi-source app, the description might need an update.

File: cmd/util/app.go [hunk 5]
```diff
@@ -751,6 +753,8 @@ func ConstructSource(source *argoappv1.ApplicationSource, appOpts AppOptions, fl
 			setPluginOptEnvs(source, appOpts.pluginEnvs)
 		case "ref":
 			source.Ref = appOpts.ref
+		case "source-name":
+			source.Name = appOpts.SourceName
```
Comment (ishitasequeira) [inline cmd/util/app.go:757]: We should be checking if the name is already assigned to some other source within the `app.spec.sources` before allowing to set it.

**Review by ishitasequeira — CHANGES_REQUESTED** [review]
@CefBoud Thanks for taking this up!! Left some comments. Also, I think this change would also need to be reflected on the Application Create and AddSource UI Pages.

File: cmd/util/app.go (hunk 1 above)
Comment (CefBoud) [reply cmd/util/app.go:92]: @agaudreault Thank you for the review. Other fields in this struct, such as `Validate` and `Parameters`, are capitalized because they are used in the `commands` package. The same applies to the newly added `SourceName`.

File: cmd/util/app.go (hunk 5 above)
Comment (CefBoud) [reply cmd/util/app.go:757]: I added the check in `validateAndNormalizeApp` to have a central validation for the both the UI and CLI API calls.

File: pkg/apis/application/v1alpha1/types.go [hunk 6]
```diff
@@ -191,6 +191,8 @@ type ApplicationSource struct {
 	Chart string `json:"chart,omitempty" protobuf:"bytes,12,opt,name=chart"`
 	// Ref is reference to another source within sources field. This field will not be used if used with a `source` tag.
 	Ref string `json:"ref,omitempty" protobuf:"bytes,13,opt,name=ref"`
+	// Name is used to refer to a source and is displayed in the UI.
```
Comment (ishitasequeira) [inline pkg/apis/application/v1alpha1/types.go:194]: Considering we want this to be of use only for a multi-source app, can we update the comment to reflect the same?

File: cmd/argocd/commands/app.go (hunk 3 above)
Comment (ishitasequeira) [reply cmd/argocd/commands/app.go:399]: That works for me. Although, let's update the comment on the `Name` field in types.go to reflect that it is mainly used for multi-source apps.

File: pkg/apis/application/v1alpha1/types.go (hunk 6 above)
Comment (CefBoud) [reply pkg/apis/application/v1alpha1/types.go]: Sure thing. Updated.

**Review by ishitasequeira — APPROVED** [review]
The PR looks good!! Thanks @CefBoud!!

**Comment by ishitasequeira** [comment]
@agaudreault do you still have any concerns on the PR?

File: cmd/argocd/commands/app.go [hunk 7]
```diff
@@ -316,6 +316,17 @@ func printHeader(acdClient argocdclient.Client, app *argoappv1.Application, ctx
 	}
 }
 
+// getSourceNameToPositionMap returns a map of source name to position
+func getSourceNameToPositionMap(app *argoappv1.Application) map[string]int64 {
+	sourceNameToPosition := make(map[string]int64)
+	for i, s := range app.Spec.Sources {
+		if s.Name != "" {
+			sourceNameToPosition[s.Name] = int64(i + 1)
```
Comment (andrii-korotkov-verkada) [inline cmd/argocd/commands/app.go:324]: It's a bit confusing that we use indexes starting from 1, but this is out of scope of this PR.

File: cmd/argocd/commands/app.go [hunk 8]
```diff
@@ -381,6 +396,19 @@ func NewApplicationGetCommand(clientOpts *argocdclient.ClientOptions) *cobra.Com
...
+			if sourceName != "" && sourcePosition != -1 {
+				errors.CheckError(fmt.Errorf("Only one of source-position and source-name can be specified."))
+			}
+
+			if sourceName != "" {
+				sourceNameToPosition := getSourceNameToPositionMap(app)
+				if pos, ok := sourceNameToPosition[sourceName]; !ok {
+					log.Fatalf("Unknown source name '%s'", sourceName)
+				} else {
+					sourcePosition = int(pos)
```
Comment (andrii-korotkov-verkada) [inline cmd/argocd/commands/app.go:408]: Why convert to int instead of using int in `getSourceNameToPositionMap`?

File: cmd/argocd/commands/app.go [hunk 9]
```diff
@@ -796,6 +829,20 @@ func NewApplicationSetCommand(clientOpts *argocdclient.ClientOptions) *cobra.Com
...
+				}
+			}
```
Comment (andrii-korotkov-verkada) [inline cmd/argocd/commands/app.go:844]: Create a helper function for getting the position, since this code is copy-pasted between multiple commands.

File: cmd/argocd/commands/app.go [hunk 10]
```diff
@@ -891,6 +942,20 @@ func NewApplicationUnsetCommand(clientOpts *argocdclient.ClientOptions) *cobra.C
...
```
Comment (andrii-korotkov-verkada) [inline cmd/argocd/commands/app.go:957]: Use a helper function.

File: ui/src/app/applications/components/application-parameters/source-panel.tsx [hunk 11]
```diff
@@ -204,6 +205,11 @@ export const SourcePanel = (props: {
                                                     </div>
                                                 </div>
                                             </div>
+                                            <div className='row argo-form-row'>
+                                                <div className='columns small-10'>
+                                                    <FormField formApi={api} label='Name' field={'spec.source.name'} component={Text}></FormField>
```
Comment (andrii-korotkov-verkada) [inline ui/src/app/applications/components/application-parameters/source-panel.tsx:210]: Should we do an optional param checks, since source and name may not be present?

File: cmd/argocd/commands/app.go (hunk 8 above)
Comment (CefBoud) [reply cmd/argocd/commands/app.go:408]: `int64`  is already being used for sourcePosition.

**Review by agaudreault — COMMENTED** [review]
LGTM. Just missing tests on the errors cases for the new parameter

File: ui/src/app/applications/components/application-parameters/source-panel.tsx (hunk 11 above)
Comment (CefBoud) [reply ui/src/app/applications/components/application-parameters/source-panel.tsx:210]: Their values are initialized with an empty string.

File: cmd/argocd/commands/app.go (hunk 9 above)
Comment (CefBoud) [reply cmd/argocd/commands/app.go:844]: The returned map is sometimes used in a loop rather than directly in an if statement. Its usage varies, especially when it’s reused within loops. Using a function would also require checking if the name exists, which means an additional if statement would be needed with the function. As a result, I'm not sure it would offer much more conciseness.

**Comment by CefBoud** [comment]
Thank you for the reviews. @agaudreault  @ishitasequeira  @andrii-korotkov-verkada 

@agaudreault I added tests to check for param errors.

File: cmd/argocd/commands/app.go (hunk 7 above)
Comment (ishitasequeira) [reply cmd/argocd/commands/app.go:325]: @andrii-korotkov-verkada , We do not consider this to be an index, but the positions. There was a discussion in the community around this and we agreed to start with 1 on CLI and 0 on apis.

Postion = 1 --> Index = 0.

**Review by agaudreault — APPROVED** [review]

**Review by agaudreault — APPROVED** [review]
